src/implementation.cpp

```diff
@@ -5,37 +5,54 @@
 
 namespace ada {
 
-  ada_warn_unused url parse(std::string input,
+  ada_warn_unused std::string to_string(ada::encoding_type type) {
+    switch(type) {
+    case ada::encoding_type::UTF8 : return "UTF-8";
+    case ada::encoding_type::UTF_16LE : return "UTF-16LE";
+    case ada::encoding_type::UTF_16BE : return "UTF-16BE";
+    default: unreachable();
+    }
+  }
+
+  ada_warn_unused url parse(std::string_view input,
                             std::optional<ada::url> base_url,
                             ada::encoding_type encoding,
                             std::optional<ada::state> state) noexcept {
-
-    return ada::parser::parse_url(std::move(input), std::move(base_url), encoding, std::nullopt, state);
-  }
-
-  ada_warn_unused url parse(std::string_view input) noexcept {
-    return ada::parser::parse_url(std::string(input));
+    // TODO std::move(base_url) might be unwise. Check.
+    return ada::parser::parse_url(input, std::move(base_url), encoding, std::nullopt, state);
   }
 
   /*
    * The protocol setter steps are to basic URL parse the given value, followed by U+003A (:),
    * with this’s URL as url and scheme start state as state override.
    *
+   * TODO: This should probably a method in the struct ada::url.
```

Comment:
I agree. I had to stop in the middle of my pull request. Appreciate if you can move this.